File: number-of-islands.py
class Solution:
    # DFS without visited: O(mn), O(1)
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        stack = []
        answer = 0
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    stack.append((i, j))
                    answer += 1
                    while stack:
                        print(stack)
                        x, y = stack.pop()
                        if grid[x][y] == '1':
                            grid[x][y] = '0'
                            if x+1 < m and grid[x+1][y] == '1':
                                stack.append((x+1, y))
                            if y+1 < n and grid[x][y+1] == '1':
                                stack.append((x, y+1))
                            if 0 <= x-1 and grid[x-1][y] == '1':
                                stack.append((x-1, y))
                            if 0 <= y-1 and grid[x][y-1] == '1':
                                stack.append((x, y-1))
        return answer
    
    # DFS: O(mn), O(mn)
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        visited = [[0 for _ in range(n)] for _ in range(m)]
        stack = []
        answer = 0
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' and not visited[i][j]:
                    stack.append((i, j))
                    answer += 1
                    while stack:
                        x, y = stack.pop()
                        if not visited[x][y]:
                            visited[x][y] = 1
                            if x+1 < m and grid[x+1][y] == '1' and not visited[x+1][y]:
                                stack.append((x+1, y))
                            if y+1 < n and grid[x][y+1] == '1' and not visited[x][y+1]:
                                stack.append((x, y+1))
                            if 0 <= x-1 and grid[x-1][y] == '1' and not visited[x-1][y]:
                                stack.append((x-1, y))
                            if 0 <= y-1 and grid[x][y-1] == '1' and not visited[x][y-1]:
                                stack.append((x, y-1))
        return answer

File: number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int M = grid.length, N = grid[0].length;
        boolean[][] visited = new boolean[M][N];
        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};
        int count = 0;

        for(int i=0; i<M; i++) {
            for(int j=0; j<N; j++) {
                if(grid[i][j] == '1' && !visited[i][j]) {
                    count++;
                    Deque<Pair> stack = new ArrayDeque<>();
                    stack.push(new Pair(i, j));
                    while(stack.size() > 0) {
                        Pair pair = stack.pop();
                        int x = pair.x, y = pair.y;
                        if(!visited[x][y]) {
                            visited[x][y] = true;
                            for(int d=0; d<4; d++) {
                                int nx = x + dx[d], ny = y + dy[d];
                                if(0 <= nx && nx < M && 0 <= ny && ny < N && !visited[nx][ny] && grid[nx][ny] == '1')
                                    stack.push(new Pair(nx, ny));
                            }
                        }
                    }
                }
            }
        }
        return count;
    }
}

class Pair {
    int x;
    int y;

    public Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}